### PR TITLE
Feature/conveyor2.0

### PR DESCRIPTION
**Pull request by toxicoreos**
PTAL

### CODE REVIEW

File: src/main/include/subsystems/Conveyor.h [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+   * Will be called periodically whenever the CommandScheduler runs.
+   */
+  void Periodic();
+
+  bool IsBallComing();
+  bool IsBallExiting();
+  void BallIntakeIncoming();
+  void BallIntakeExiting();
+  void SetConveyor();
+  int CurrentBallCount();
```
Comment (duane701) [inline src/main/include/subsystems/Conveyor.h]: Let's rename to `BallCount`. The count is always current.

```suggestion
  int BallCount();
```

File: src/main/cpp/subsystems/Conveyor.cpp [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+void Conveyor::BallIntakeIncoming()
+{
+    ++mBallCount;
+}
+void Conveyor::BallIntakeExiting()
+{
+    --mBallCount;
+}
+void Conveyor::SetConveyor()
```
Comment (duane701) [inline src/main/cpp/subsystems/Conveyor.cpp]: `SetConveyor` looks like a command. Should this logic be in a command?

File: src/main/include/commands/SetConveyor.h [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+#include <frc2/command/CommandHelper.h>
+#include <subsystems/Conveyor.h>
+
+/**
+ * An example command.
+ *
+ * <p>Note that this extends CommandHelper, rather extending CommandBase
+ * directly; this is crucially important, or else the decorator functions in
+ * Command will *not* work!
+ */
```
Comment (duane701) [inline src/main/include/commands/SetConveyor.h:20]: This comment can be removed, next time this file is edited.

File: src/main/cpp/commands/SetConveyor.cpp [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  AddRequirements(&mConveyor);
+}
+
+// Called when the command is initially scheduled.
+void SetConveyor::Initialize() {}
+
+// Called repeatedly when this Command is scheduled to run
+void SetConveyor::Execute() 
+{
+  mConveyor.SetConveyor();
```
Comment (duane701) [inline src/main/cpp/commands/SetConveyor.cpp]: The code in `Conveyor::SetConveyor` should be here. `Conveyor::SetConveyor` should be much more simple.

File: src/main/cpp/RobotContainer.cpp [hunk 5]
```diff
@@ -21,12 +22,16 @@ RobotContainer::RobotContainer()
...
+  mConveyor.SetDefaultCommand(SetConveyor(mConveyor));
+
   // Configure the button bindings
   ConfigureButtonBindings();
 }
 
 void RobotContainer::ConfigureButtonBindings() {
   // Configure your button bindings here
+  frc2::Trigger( [this] { return mConveyor.IsBallComing(); }).WhenActive( [this]{ mConveyor.BallIntakeIncoming(); });
+  frc2::Trigger( [this] { return mConveyor.IsBallExiting(); }).WhenInactive( [this] { mConveyor.BallIntakeExiting(); });
```
Comment (duane701) [inline src/main/cpp/RobotContainer.cpp:34]: I think we need the complements of these buttons as well. That is we need both Active and Inactive states from both sensors.

A logic diagram for ball counting should be designed. Some test code would be useful.

**Review by duane701 — CHANGES_REQUESTED** [review]

File: src/main/cpp/subsystems/Conveyor.cpp (hunk 2 above)
Comment (duane701) [reply src/main/cpp/subsystems/Conveyor.cpp]: I'm open to an argument that we need to leave this in place until we design a complete ball counter. But ... you need to convince me.

File: src/main/cpp/subsystems/Conveyor.cpp [hunk 6]
```diff
@@ -39,22 +39,18 @@ void Conveyor::BallIntakeExiting()
 }
 void Conveyor::SetConveyor()
 {
-    if(CurrentBallCount() == 0 && !IsBallComing() && IsBallExiting())
+    if((IsBallComing() && BallCount() < 5) /* TODO: add in here: || (BallCount() > 0 && IsShooting()) */ )
```
Comment (duane701) [inline src/main/cpp/subsystems/Conveyor.cpp]: Add an `IsShooting()` method to Conveyor that returns `false`.  With the todo in that function to calling the shooter.

**Comment by NWSchallock** [comment]
PTAL @duane701

File: src/main/include/RobotContainer.h [hunk 7]
```diff
@@ -12,7 +12,8 @@
 #include <frc/XboxController.h>
 #include <frc2/command/Command.h>
 
-#include "subsystems/Chassis.h"
+#include <subsystems/Chassis.h>
+#include <subsystems/Conveyor.h>
```
Comment (duane701) [inline src/main/include/RobotContainer.h:16]: Our subsystems use `"`

```suggestion
#include "subsystems/Conveyor.h"
```

**Review by duane701 — CHANGES_REQUESTED** [review]
FTP

**Comment by duane701** [comment]
LGTM -- Rebase after Jason

Or you can race, first one wins, last on rebases again, and again, and again.

**Comment by NWSchallock** [comment]
Replaced by #33

**Pull request closed without merging**